instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.javaagent.runtimemetrics.java8;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.AttributesBuilder;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Properties;
+import java.util.jar.JarEntry;
+import java.util.jar.JarInputStream;
+import java.util.jar.Manifest;
+
+final class JarAnalyzerUtil {
+
+  static final AttributeKey<String> PACKAGE_NAME = AttributeKey.stringKey("package.name");
+  static final AttributeKey<String> PACKAGE_VERSION = AttributeKey.stringKey("package.version");
+  static final AttributeKey<String> PACKAGE_TYPE = AttributeKey.stringKey("package.type");
+  static final AttributeKey<String> PACKAGE_DESCRIPTION =
+      AttributeKey.stringKey("package.description");
+  static final AttributeKey<String> PACKAGE_CHECKSUM = AttributeKey.stringKey("package.checksum");
+  static final AttributeKey<String> PACKAGE_PATH = AttributeKey.stringKey("package.path");
+
+  /**
+   * Set the attributes {@link #PACKAGE_TYPE} from the {@code archiveUrl}.
+   *
+   * <p>The {@link #PACKAGE_TYPE} is set extension of the archive, e.g. {@code jar}.
+   */
+  static void addPackageType(AttributesBuilder builder, URL archiveUrl) throws Exception {
+    String path = archiveUrl.getFile();
+    int extensionStart = path.lastIndexOf(".");
+    if (extensionStart > -1) {
+      builder.put(PACKAGE_TYPE, path.substring(extensionStart + 1));
+      return;
+    }
+    throw new Exception("Cannot extract archive type from URL: " + archiveUrl);
+  }
+
+  /**
+   * Set the attributes {@link #PACKAGE_CHECKSUM} from the {@code archiveUrl}.
+   *
+   * <p>The {@link #PACKAGE_CHECKSUM} is set to the SHA-1 checksum of the archive, e.g. {@code
+   * 30d16ec2aef6d8094c5e2dce1d95034ca8b6cb42}.
+   */
+  static void addPackageChecksum(AttributesBuilder builder, URL archiveUrl) throws IOException {
+    builder.put(PACKAGE_CHECKSUM, computeSha1(archiveUrl));
+  }
+
+  private static String computeSha1(URL archiveUrl) throws IOException {
+    MessageDigest md;
+    try {
+      md = MessageDigest.getInstance("SHA1");
```

Comment:
Is `MessageDigest` reusable? Could we perhaps use a single instance of that?

Reply:
`MessageDigest` has a `reset` method

Reply:
Would have to take a lock on it to make sure its safe, but that's not a problem because as of now the processing of jars is single threaded.

Reply:
A `ThreadLocal` would work fine as well, I think